core/src/main/java/jenkins/scm/RunWithSCM.java

```diff
@@ -0,0 +1,160 @@
+/*
+ * The MIT License
+ *
+ * Copyright 2017 CloudBees, inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package jenkins.scm;
+
+import com.google.common.collect.ImmutableSet;
+import hudson.model.Job;
+import hudson.model.Queue;
+import hudson.model.Result;
+import hudson.model.Run;
+import hudson.model.User;
+import hudson.scm.ChangeLogSet;
+import hudson.scm.SCM;
+import hudson.util.AdaptedIterator;
+import org.kohsuke.stapler.export.Exported;
+
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
+import java.util.AbstractSet;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Set;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * Allows a {@link Run} to provide {@link SCM}-related methods, such as providing changesets and culprits.
+ *
+ * @since FIXME
+ */
+public interface RunWithSCM<JobT extends Job<JobT, RunT> & Queue.Task,
```

Comment:
I am not sure it is safe to use Queue.Task/Executable here. It may be possible to imagine other Run types, and it does not seem to be really required. 🐜 , will be bug if I find a real use-case for it